### PR TITLE
[SPARK-32064][SQL] Supporting create temporary table

### PR DESCRIPTION
**Pull request by LantaoJin**
### What changes were proposed in this pull request?

Many databases and data warehouse SQL engines support temporary tables. A temporary table, as its named implied, is a short-lived table that its life will be only for current session.
[Hive Temporary Table](https://docs.cloudera.com/HDPDocuments/HDP3/HDP-3.0.1/using-hiveql/content/hive_create_a_hive_temporary_table.html)
[Teradata Volatile Table](https://docs.teradata.com/reader/rgAb27O_xRmMVc_aQq2VGw/mpJF1z_vSlpMbZYxFmRJfA)
[PostgreSQL Temporary Table](https://www.postgresql.org/docs/12/sql-createtable.html)

In Spark, there is no temporary table. the DDL “CREATE TEMPORARY TABLE AS SELECT” will create a temporary view. A temporary view is totally different with a temporary table. 

This ticket to support Spark native temporary table. More details are described in [DESIGN DOCS](https://docs.google.com/document/d/1RS4Q3VbxlZ_Yy0fdWgTJ-k0QxFd1dToCqpLAYvIJ34U/edit?usp=sharing)

Parent ticket https://issues.apache.org/jira/browse/SPARK-32063

### Why are the changes needed?

A temporary view is just a VIEW. It doesn’t materialize data in storage. So it has below shortage:

1. View will not give improved performance. Materialize intermediate data in temporary tables for a complex query will accurate queries, especially in an ETL pipeline.
2. View which calls other views can cause severe performance issues. Even, executing a very complex view may fail in Spark. 
3. Temporary view has no database namespace. In some complex ETL pipelines or data warehouse applications, without database prefix is not convenient. It needs some tables which only used in current session.

### Does this PR introduce _any_ user-facing change?
YES.
```sql
CREATE TEMPORARY TABLE tt1 AS SELECT ..
```
before the patch, it will create a local temporary VIEW. After this patch, it will create a temporary table.
```sql
CREATE TEMPORARY TABLE tt1 USING ..
```
before the patch, it will throw exception. After this patch, it will create a temporary table.

Add a new API in `Catalog.scala`
```scala
def dropTempTable(tableName: String): Boolean
```

### How was this patch tested?
Add unit tests.

### CODE REVIEW

**Comment by LantaoJin** [comment]
retest this please

**Comment by LantaoJin** [comment]
retest this please

**Comment by LantaoJin** [comment]
retest this please

**Comment by LantaoJin** [comment]
retest this please

**Review by gatorsmile — COMMENTED** [review]
This feature requires a lot of changes in different places. We need to define whether it should be global or local; whether we should create such a schema in each session; various error handling when the expected table dropping is not completely finished. 

Trying to understand your use case first. Instead of creating a regular table, you want to create a temp table that does not need to be manually dropped?

**Comment by LantaoJin** [comment]
@gatorsmile Yes. Just like Hive temporary table or Teradata volatile table. We are migrating our Spark to v3.0. This is one of inside features which had widely used in our prodution.

**Comment by LantaoJin** [comment]
retest this please

**Comment by LantaoJin** [comment]
@gatorsmile @cloud-fan Current implementation is not complex. Any comments?

**Comment by gatorsmile** [comment]
For a proper support, this requires more discussions about the semantics. 

Also, we need to list the expected behaviors for all the statements listed in https://spark.apache.org/docs/latest/sql-ref-syntax.html . 

So far, this PR and design doc does not have the corresponding contents.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 1]
```diff
@@ -3030,6 +3031,215 @@ abstract class DDLSuite extends QueryTest with SQLTestUtils {
       }
     }
   }
+
+  test("create temporary table using data source") {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:3035]: maybe create a new suite for these?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala [hunk 2]
```diff
@@ -226,4 +227,22 @@ object StaticSQLConf {
       .version("3.0.0")
       .intConf
       .createWithDefault(100)
+
+  val SPARK_SCRATCH_DIR =
+    buildStaticConf("spark.scratchdir")
+      .doc("Scratch space for Spark temporary table and so on. Similar with hive.exec.scratchdir")
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala:233]: let's not bring up hive here. Slowly nobody will care about Hive.

Also this should be spark.sql.scratchdir?

**Comment by LantaoJin** [comment]
> For a proper support, this requires more discussions about the semantics.
> 
> Also, we need to list the expected behaviors for all the statements listed in https://spark.apache.org/docs/latest/sql-ref-syntax.html .
> 
> So far, this PR and design doc does not have the corresponding contents.

Sure, if you could tell me more about what do we need to discuss and what details should be written in documentation, that would be very helpful to me. About the concept of "temporary table", I think it is widly used in database domain: MySQL, PostgreSQL, Oracle etc, also in data warehouse domain: Hive, Teradata etc. Even though their implementations and grammar maybe different more or less, the purposes are similar in my opinion. This implementation and grammar of `Spark temporary table` mainly references [Hive](https://docs.cloudera.com/HDPDocuments/HDP3/HDP-3.0.1/using-hiveql/content/hive_create_a_hive_temporary_table.html) and [PostgreSQL](https://www.postgresql.org/docs/12/sql-createtable.html). They are basicly same.

**Comment by cloud-fan** [comment]
If I write the output to a temp location and then create a temp view, is it similar to the temp table? Except that temp table can be removed when the session terminates.

**Comment by LantaoJin** [comment]
> If I write the output to a temp location and then create a temp view, is it similar to the temp table? Except that temp table can be removed when the session terminates.

There is no path and materialized data for temp view. So the answer is no. You can simply treate a `temporary table` as a Spark permanent data source table which will be dropped automatically when the session closed. So the implementation is not complex. But the user cases of temporary table are more than it looks. A permanent metastore table needs more maintenance, creating a permanent metastore table needs write permission of database it related and folder permission of storage. For many ad-hoc user case (OLAP), a user may only have limited permission like read. So current, users use `temporary view` to implement their complicated queries. View will not give improved performance without data materialzation. So if user can create `temporary table`, no need to grant write permission in production databases, it is very convenience for users.

Imaging the user case like below:
A databricks runtime user login to the notebook. He/She will write some statements to do an analysis job. Maybe in databricks  runtime, user has r/w permissions on his/her own default space (we call it workspace), but no write permission on production database (for example, database "dw"). Without `temporary table`, they may use temporary view in their SQL statements. Or they can create a temporary workspace/database (for example, database named "tony_work"), and create permanent tables in "tony_work" then drop them all when they logout (if they can logout without failure). But users may want to share their scripts (above SQL statements) to another user or batch account. He/She has to change their scripts since the batch account or another user don't have the write permission on database "tony_ work". So in our production, the `temporary table` feature is wildly used, especially for the Teradata users who migrated to Spark. @cloud-fan

**Comment by cloud-fan** [comment]
Let me clarify my previous comment a little bit: If I write the output to a temp location and then create a temp view **to read from this temp location**, ...

I agree with your use case, but I'm bit worried about adding a new big API (temp table) if there are easy workarounds.

**Comment by LantaoJin** [comment]
> If I write the output to a temp location and then create a temp view to read from this temp location, ...

Ah, I knew your meaning now, using `CREATE TEMP VIEW USING` command. I think it's still tricky. First, platform needs to provide the "temp" location to storage users' data which are able to clean all untracked files. It's hard to do that. Second, "CREATE TEMP VIEW USING" command lack the view definition. In most cases, immutable `temporary table` is enough. After the temporary created with `AS SELECT ...`, less insert/overwrite operations will perform on it. `CREATE TEMP VIEW USING` command without `AS SELECT`, a good practise seems to `insert/overwrite` it then. What's more, the storage folder permission is still a problem in a ACL enabled platform (even Spark code didn't import ACL, many companies build their platforms similar with Databricks Runtime)

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala (hunk 1 above)
Comment (LantaoJin) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:3035]: sure

**Comment by LantaoJin** [comment]
retest this please

**Comment by LantaoJin** [comment]
retest this please

**Comment by LantaoJin** [comment]
retest this please

**Comment by kevinjmh** [comment]
how about to use `CACHE TABLE` command to do that?

**Comment by LantaoJin** [comment]
> how about to use `CACHE TABLE` command to do that?

I think you mean `CACHE VIEW` since `CACHE TABLE` still needs to `DROP TABLE` manually.
There are four reasons to build `temporary table` instead of `cache a temporary view`:
1. The intermediate table which users want to create as a temporary table are always very large. To avoid OOM, user has to use `CACHE TABLE viewname OPTIONS('storageLevel', 'disk_only')`. It's not friend to SQL users. Users confuse what 'storageLevel' and 'disk_only' are.
2. View is dynamic and table is static. Whatever the underly detail tables changes, accessing a view should always access the latest data. So when the underly detail tables of a view changed, Spark will recache all data to executors's local disks again. For a large intermediate table, this is not performance friendly.
3. The storages between cached view and temporary table are different. Cache command stores the block in executors local disk which managed by `blockManager`, and data of temporary table is stored in external storage like HDFS. Local disks in executors are very limited and not easy to scale out. Besides, the data in HDFS can be organized by Parquet file format, this can highly benefits `Scan` operation and predicates pushdown.
4. The accuracy of table statistics for cached view is easily expired. IIUC, table statistics for cached view are calculated when the cache operation occurs. When the data of under details tables changed, the statistics of a cached view won't be updated. So some optimization like AQE cannot work correctly. But to a temporary table, no problem.

**Pull request closed without merging**